Kubelet pod eviction proposal

derekwaynecarr opened the pull request:
The following is a proposal for how the `kubelet` may pro-actively fail a pod in response to local compute resources being starved.  The proposal focuses on memory as a first candidate, and defines a `greedy` strategy for reclaiming starved resources on the node since it seemed easiest to describe for operators versus other options and probably satisfies a broad set of use case environments.

Putting this out now for community feedback, but anticipate some more refinement around how we report eviction configuration back to users in the `Node API`.

/cc @bgrant0607 @smarterclayton @vishh @dchen1107 @kubernetes/rh-cluster-infra @kubernetes/goog-node

derekwaynecarr committed:
Kubelet pod eviction proposal

pmorie commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+## Scope of proposal
+
+This proposal defines a pod eviction policy for reclaiming compute resources.
```
...and preventing out of resource situations.  ?

timothysc replied:
That's taints.

pmorie commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+In the first iteration, it focuses on memory; later iterations are expected to cover
+other resources like disk.  The proposal focuses on a simple default core policy
+intended to cover the broadest class of user workloads.
```
Are the mechanics intented to be portable to future resources?

pmorie replied:
Nvm, I continued reading

pmorie commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+The `kubelet` will accept a flag to define the eviction threshold as follows:
+
+* `--eviction-threshold=:` - A set of `ResourceName`=`ResourceQuantity` pairs that
```
how are they delimited?

derekwaynecarr replied:
comma-delimited similar to `system-reserved` and `kube-reserved` flags

pmorie commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+### Limit total amount of resources for BestEffort pods
+
+The `kubelet` may want to limit the total amount of resources available for
+BestEffort pods to induce their eviction even if the eviction-threshold has
```
you monospace BestEffort everywhere else.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+In the first iteration, it focuses on memory; later iterations are expected to cover
+other resources like disk.  The proposal focuses on a simple default core policy
+intended to cover the broadest class of user workloads.
+
```
Can we clarify the higher level requirements or goals explicitly before proposing solutions?

derekwaynecarr replied:
Sure, will add a section on goals.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+The node will only support `memory` first, but other resources such as `cpu` or `disk` may be
+added in the future.
+
+If no eviction threshold is specified for a compute resource, the node may choose a sane
```
This suggests that users have to set `--eviction-threshold=cpu=0,disk=0` by default. Why?
Why not default to no-op unless kubelet is configured to evict pods?

derekwaynecarr replied:
At the time of authoring this, I was trying to be consistent with what is in the Node Allocable proposal where it suggested a sane default.  I am fone defaulting to no-op if unspecified.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+The `kubelet` will accept a flag to define the eviction monitoring interval as follows:
+
+* `--eviction-monitoring-interval=:` - Interval at which the kubelet monitors available
```
Why does this have to be configurable?

vishh replied:
We have defaults set for cAdvisor housekeeping through detailed analysis. Adding a new flag might complicate internals.

derekwaynecarr replied:
I am fine with sticking with the cAdvisor defaults.  Thanks for the heads up.  One less thing :-)

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+The `kubelet` will accept a flag to define the eviction window as follows:
+
+* `--eviction-window=:` - Duration of time where the node is able to exceed the
```
So if `--eviction-threshold=memory=256Mi` and `--eviction-window=5s`, a system OOM is essentially possible even with this design?
Should we instead have soft and hard limits? Exceeding soft limits for `eviction_grace_period` is acceptible, but exceeding `hard_limit` is not acceptable?

derekwaynecarr replied:
I wasn't sure if we wanted to design for hard and soft limits in the first pass.  I have no objection to them in concept.  To be clear, system OOMs are always possible within a window of time.   If we default to cAdvisor 10s measurement interval, then we may want to define a soft limit for longer duration and a hard limit for immediate action.  Will update with this feedback in mind.

timothysc replied:
I always have a problem with soft limits.  Do you really want to evict in the case of a soft limit for a period of time?  If it's not causing harm then why care.

derekwaynecarr replied:
@timothysc - at F2F we decided to allow for soft limits, so per your concern, the net is it will just be an operator choice to use it or not.  If they want to ignore it, they can just set a hard limit and know it will be enforced during feasibility checks prior to starting a new container, and it will be validated every 10s to cause immediate eviction.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+* `--eviction-window=30s`
+* `--eviction-monitoring-interval=10s`
+
+Then the `kubelet` will interact with `cAdvisor` every `10s` to introspect
```
I'd like kubelet to have one internal metrics module that is shared for all functionalities. So shall we avoid including metrics polling duration here?

derekwaynecarr replied:
Yep

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+If a compute resource has reached its eviction threshold over the `30s` period
+the `kubelet` will initiate eviction to attempt to reclaim the resource that
+has met its eviction threshold.  The `kubelet` will record an event when eviction
+is initiated on the node.
```
What information is contained in the event?

derekwaynecarr replied:
I will add a sample event with reason

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+It ranks pods within a quality of service tier as follows:
+
+* `BestEffort` pods that consume the most of the starved resource are failed
```
Specifically within `BestEffort` I'd like to choose the most-recent jobs first.

derekwaynecarr replied:
I went back and forth on that and thought keeping greedy consistent made sense to me across all qos tiers since it seemed most in line with our existing OOM killer process.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+relative to their request for that resource are killed first.  If no pod
+has exceeded its request, the strategy targets the largest consumer of the
+starved resource.
+* `Guaranteed` pods that consume the greatest amount of the starved resource
```
Why would we ever have to evict Guaranteed pods? The scheduler essentially failed if we have to evict `Guaranteed` pods. Can we start with not Evicting `Guaranteed` pods at all?

derekwaynecarr replied:
See comment above on burstable.

vishh replied:
I don't think it makes sense to ever evict `Guaranteed` pods.

timothysc replied:
> I don't think it makes sense to ever evict Guaranteed pods.

+1

vishh replied:
@derekwaynecarr I still think that it does not make sense to evict Guaranteed pods and even Burstable pods if their usage is < request. It violates QoS.

vishh replied:
Ping @derekwaynecarr

derekwaynecarr replied:
see text below

derekwaynecarr replied:
QoS is only as good as the foundation its built upon.  A guaranteed pod is guaranteed to never be evicted because of another pod.  If the node is on fire because the docker daemon or the kubelet or some other system daemon is exceeding their expected compute budgets, the guarantee is worth nothing.  If two guaranteed pods are on the same node, do you want to kill the experience for both pods by not evicting one of them and having the entire system OOM?

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+* `BestEffort` pods that consume the most of the starved resource are failed
+first
+* `Burstable` pods that consume the greatest amount of the starved resource
```
If no pods in the `Burstable` class are exceeding their request, why will we have to evict any pods?
Is it because the node overhead became too high?
Eviction of pods with resource requests isn't ideal.

derekwaynecarr replied:
If we are not enforcing system reserved and kube reserved than either of those groups could induce memory pressure and cause a need to evict Burstable.

If we had memory soft limits in place we could hope to avoid this a little better, but I don't think we can say it will never ever happen until we make a more explicit statement on enforcing the limits on system and kube daemons.  That said, I think their enforcement if done should be an operator decision.  Operator may choose for example to let docker memory growth grow and not kill that to avoid all containers from restarting.  At least that was my thought process.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+It is anticipated other eviction strategies may be used in the future.
+
+For example, operators may want to target `BestEffort` pods that were most
+recently scheduled on the node instead of targeting `BestEffort` pods that
```
Why not both? Apply different weights for age and resource consumption.

derekwaynecarr replied:
I wasn't sure how I wanted to model weights across various axis points which is why I went with a single flag around a greedy strategy.  I am fine with a weighted system.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+to reclaim enough resource on the node.  The idea being that those low resource
+pods could fit elsewhere on the cluster.
+
+Finally, operators may want to target pods based on other criteria altogether.
```
Or data locality.. If the pod has to run on the same node, its better to not evict it.
Example DB's that use local SSDs.

derekwaynecarr replied:
The list was intentionally not exhaustive, but will add more potential scenarios to describe future enhancement opportunities :-)

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+### Feasibility checks during kubelet admission
+
+The `kubelet` will reject `BestEffort` pods if the current availability is
```
+1

timothysc replied:
xref: https://github.com/kubernetes/kubernetes/pull/18263

derekwaynecarr replied:
@timothysc - once taints are defined, we will fill a value.  we discussed at F2F that we would have a new node condition that would be reported in the interim that parallels "OutOfDisk", so if memory is low, we would have a "OutOfMemory" node condition reported.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+The `kubelet` supports a flag to enable this behavior:
+
+* `--eviction-on-admission=true` - Do not start and fail BestEffort pods if 
```
Do we really need a flag?

derekwaynecarr replied:
I don't intend to ever run with the flag set to false, but put a flag there since I thought @davidopp expressed some reservation on the Kubelet doing feasibility checks until the scheduler had more back off capability.

The other reason would be if the operator did not intend to evict on most recently scheduled since this basically gives guaranteed lowest priority to that class of best effort pod.

derekwaynecarr replied:
I am removing this flag, and will always perform feasibility checks.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+The `kubelet` supports a flag to enable this behavior as follows:
+
+* `--limit-best-effort=:` A set of `ResourceName`=`ResourceQuantity` pairs that
```
Why do we need this feature? A pod of higher class should be able to evict BestEffort pods.

derekwaynecarr replied:
I agree that a pod of higher class should evict.

This was in response to your point 3 on this issue.

https://github.com/kubernetes/kubernetes/issues/18438

It's possible I misunderstood what you meant by a cap and if it should run separate from eviction or not require it to define a cap.

derekwaynecarr replied:
I am removing this feature.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+given monitoring interval and eviction-window before the `kubelet` is able to fail
+pods and reclaim starved resource.
+
+The `kubelet` in response to a system OOM event should apply back-pressure on
```
Keeping the pods pending on the node is probably going to complicate the system.
Since the rest of the features in this proposal are essentially trying to limit systemOOMs, should we try to punt on this feature for now?

derekwaynecarr replied:
This was in response to point 1 on your comment here.  

https://github.com/kubernetes/kubernetes/issues/18438

Admittedly, I gave less design thought in this space than the other areas.  I am not even 100% sure if staggered start is something we can control since containers with a restart policy of always should just be started by docker itself if I recall.

Assuming we do feasibility checks on best effort pods during admission, and we are comfortable with potentially evicting burstable containers that request too few resources but immediately attempt to consume all of a nodes available memory, this may in fact not be needed.  In general, how do you guys handle burstable pods that request too few resources and specify no limit and then proceed to try and use all available node resources?   Absent a resizer that put a limit in place, you have to evict that container or quarantine it somehow.

derekwaynecarr replied:
agree to remove system oom backpressure.

derekwaynecarr commented:
@vish - appreciate the initial review.  I am out of office remainder of year, but will look to update by Jan 4 with any accumulated review comments.  At first glance, I have no major issues with any of the suggestions so I suspect we can get closure first week of January.

vishh commented:
SGTM. Have a great vacation!!

ncdc commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+Then the `kubelet` will interact with `cAdvisor` every `10s` to introspect
+current node usage.
+
+At each monitoring interval, if a compute resource has reached it's eviction
```
its

ncdc commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+threshold, the `kubelet` will record an event related to the node describing
+which resource is low.
+
+If a compute resource has reached its eviction threshold over the `30s` period
```
comma at EOL

ncdc commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+## Eviction of Pods
+
+If the eviction criteria have been met, the `kubelet` will evict containers
```
evict containers or pods?

derekwaynecarr replied:
Pods

davidopp commented:
I haven't had time to read the proposal, but starvation detection and killing is something we've discussed for rescheduler  (#12140). I don't think I have an objection to doing it in the kubelet, but we should give some thought about what should go in the rescheduler and what should go on the kubelet.

derekwaynecarr commented:
@timothysc - fyi

derekwaynecarr commented:
A few notes based on discussions:

1. Need to add soft and hard thresholds
2. Each threshold is a pair of a signal (resource constraint) with a grace period
3. The configuration should be moved away from arguments into a versioned config file
4. The node.status can report the thresholds (signal/graceperiod), but does not need to report policy/strategy
5. The kubelet should preserve daemon set derived pods when node when that information is available
6. The kubelet should report a node condition that is related to the signal under distress based on the signal.
7. The strategy object should be a plug-in (compiled in).

Future
The cpu load factor could be used as a future signal but not near term.

derekwaynecarr commented:
Another topic:

It's worth discussing if a DaemonSet should reject BestEffort pods, and require a Burstable/Guaranteed qos given implied eviction priority.

timothysc commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+## Motivation
+
+The node needs a mechanism to preserve stability when available compute resources are low.
```
The node needs a policy mechanism...  

where policy could be plugable, etc.

jdef replied:
+1 for plugable policy

timothysc commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+If the eviction criteria have been met, the `kubelet` will evict containers
+based on their quality of service guarantee.  Within a particular quality of
+service tier, the `kubelet` may use an eviction strategy to determine the
```
as mentioned in the discussion, policy should probably be plugable with some defaults as admins way wish to define their own policy.

timothysc commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,201 @@
+
+The `kubelet` will accept a flag to define the eviction strategy as follows:
+
+* `--eviction-strategy=:"` - The strategy to use when evicting pods
```
perfer plugin vs. hard coding strategy without an expression syntax.

derekwaynecarr commented:
A few more notes:

* If you are evicted, you are immediately killed synchornously, i.e. the gracefulTerminationSeconds on eviction is not respected so as to ensure that we can actually maintain node stability.
* Eviction decisions are synch, i.e. we identify candidate pod, we kill it, if necessary, we find next candidate, we kill it., etc.  no paralllel killing of pods.  we kill a single pod in a single monitoring interval to allow for reclamation to take place.

derekwaynecarr commented:
@vishh @dchen1107 - I finally got around to updating the document based on the outcome of our F2F last month.  I am looking to begin prototyping this effort in preparation for when kube 1.2 is cut and kube 1.3 is opened, so any feedback I may have missed and should capture is appreciated.

derekwaynecarr commented:
@bgrant0607 - do you have time to review and weigh in on this proposal before I get too far in implementation?  I think it represents what we have previously discussed.

derekwaynecarr commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,223 @@
+
+const (
+	// EvictionMemoryAvailable, Memory available, in bytes
+	EvictionMemoryAvailable EvictionSignal = "memory.available"
```
@dchen1107 @vishh - input appreciated on experience from borg/omega... do you find it more natural to define a signal based on memory.usage or what I thought of as memory.available here as basically free+buffers+cache?

I had looked to use memory.available since I thought it was more natural to configure across any machine size rather than needing to set a machine relative size which is what usage would require.

derekwaynecarr replied:
So from Slack chat, I would propose following:

1.  Define `memory.available = machineInfo.capacity.memory - node.stats.workingSetBytes`
2.  Add `availableBytes` to `MemoryStats` in summary API defined as above
3.  Let users define eviction thresholds as proposed using `memory.available < 200Mi` which is easy to reason about.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,200 @@
+The `kubelet` has some support for influencing system behavior in response to a system OOM by
+having the system OOM killer see higher OOM score adjust scores for containers that have consumed
+the largest amount of memory relative to their request.  System OOM events are very compute
+intensive, and can starve the node until the OOM killing process has completed.  In addition,
```
nit: Would `stall` be more appropriate instead of `starve`?

derekwaynecarr replied:
stall is more accurate, will update.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,200 @@
+the largest amount of memory relative to their request.  System OOM events are very compute
+intensive, and can starve the node until the OOM killing process has completed.  In addition,
+the system is prone to return to an unstable state since the containers that are killed due to OOM
+are just restarted, and the node may find itself soon starved of memory again. 
```
either restarted or a new pod is scheduled on to the node.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,200 @@
+
+Instead, we would prefer a system where the `kubelet` can pro-actively monitor for
+and prevent against total starvation of a compute resource, and in cases of where it
+could appear to occur, pro-actively fail a pod and terminate its containers so its workload
```
fail `one or more pods`

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,200 @@
+
+This proposal defines a pod eviction policy for reclaiming compute resources.
+
+In the first iteration, it focuses on memory; later iterations are expected to cover
```
Is the intention to keep the kubelet UX/CLI common across all resources?

derekwaynecarr replied:
Yes, I think I should be able to set an eviction threshold in the same manner for disk as I would memory.  I think the syntax allows for that.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,200 @@
+to reclaim the associated starved resource.
+
+Initial support for the feature will focus on available memory, but other potential signals
+that may be added in the future include cpu load or available disk.
```
AFAIK, we decided not to implement `cpu` based evictions for now. I'd prefer not mentioning it to avoid someone posting a PR for it.

derekwaynecarr replied:
Sure.

vishh commented on docs/proposals/kubelet-eviction.md:
````diff
@@ -0,0 +1,200 @@
+To configure hard eviction thresholds, the following flag will be supported:
+
+```
+--eviction-hard="": Comma-delimited list of eviction expressions.  For example, 'memory.available<300Mi'
````
Where is the schema for this flag?

derekwaynecarr replied:
Right now, I was only advocating support for `memory.available`, I can clarify.

vishh commented on docs/proposals/kubelet-eviction.md:
````diff
@@ -0,0 +1,200 @@
+```
+
+If no eviction period is configured for a corresponding soft eviction threshold,
+it will default to 30s for each soft eviction signal.
````
`30s` seems too little for disks. On the other hand, having per-resource defaults makes documentation messier. Should we just not mention the defaults instead? How would it impact the end user?

derekwaynecarr replied:
@vishh - we could just require the user to provide a value explicitly.

So the following would error:

```
--eviction-soft=memory.available<500Mi
```

because no corresponding `--eviction-soft-period` was provided.

The following would be valid:

```
--eviction-soft=memory.available<500Mi
--eviction-soft-period=memory.available=30s
```

If we configure disk:


```
--eviction-soft=memory.available<500Mi,fs.available<200Mi
--eviction-soft-period=memory.available=30s,fs.available=20s
```

etc.

vishh replied:
```
--eviction-soft=memory.available<500Mi,fs.available<200Mi
--eviction-soft-period=memory.available=30s,fs.available=20s
```
This makes sense.... Thanks for clarifying. I hope the flag defaults will be explicit.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,200 @@
+`<eviction-signal><operator><quantity>`
+
+* valid `eviction-signal` tokens are `memory.available` (future signals are anticipated `storage` or `cpu.load`)
+* valid `operator` tokens are `<` and `>`
```
Why would we need `<`?

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,200 @@
+
+The eviction sequence works as follows:
+
+* for each monitoring interval, if eviction thresholds have been exceeded
```
This interval would change based on how long it takes for the pod to get cleanup right?

derekwaynecarr replied:
Correct.

The current prototype I have does the following:

```
// called every monitoring interval
func (m *manager) synchronize() {
  stats := summary_stats
  if podsPendingEvictionStillActive(...) {
    return
  }
  if checkIfEvictionThresholdsAreMet(...) {
     podToEvict := strategy.Select(...)
     // toggle state to block incoming pods on admission, etc.
  }
}
```

So basically, we do not find a new pod to evict until we have definitively observed the old pod being evicted.  It has pros/cons, but it integrated best with current kubelet sync loop architecture.  Basically, the `getPodsToSync()``` ticker every 1s is used to query the stability manager to find the pods that need to be evicted.

Anyway, this is a little bit in the weeds, we can discuss that on the implementation PR.  Major thing is that we observe a pod eviction is completed before evicting another.

vishh replied:
LGTM

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,200 @@
+
+### Alternative eviction strategies
+
+It is anticipated other eviction strategies may be used in the future.
```
These strategies can be added to the default `greedy` policy as well, since it is generally beneficial to all users.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,200 @@
+
+### Feasibility checks during kubelet admission
+
+The `kubelet` will reject `BestEffort` pods if any of its associated
```
In case the scheduler messed up, the kubelet might have to reject `Guaranteed` pods as well.

vishh commented:
Completed review pass. It LGTM mostly. Just a few questions/comments. Very well written @derekwaynecarr !!!

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,200 @@
+
+### Eviction Strategy: greedy
+
+The `greedy` strategy targets pods that are the largest consumers of the
```
Given that we are adding `Priority` to pods, should we even support multiple policies? [Pod Priority proposal](https://github.com/kubernetes/kubernetes/pull/22217/files#diff-8141aabac2eced92d5560d2d24c1a2d7R143)

The current `greedy` policy would then evolve into a `default` policy.

davidopp replied:
I don't have any context here as I haven't read this proposal, but I just wanted to make a drive-by comment: the current plan is to not use scheduling priority for any decisions on-node. Only QoS would be a factor for on-node decisions. Please discuss on #22212 if you want to discuss that topic.

derekwaynecarr replied:
@vishh - removed references to multiple policies.  agree its simpler to just iterate on a basic policy now.

mdshuai commented:
@derekwaynecarr How to evict pod that has specified nodeName ? same with daemonset part ?

derekwaynecarr commented:
@mdshuai - a pod that has a specified node name is just a pod that is scheduled to the node.  Eviction happens at the node itself, so the pod if chosen for eviction would just have a Failed status with the eviction reason.

To support not targeting daemon set pods, we would need support for the owner reference concept here: https://github.com/kubernetes/kubernetes/pull/23656 .  This proposal does not assume the existence of that one yet, so I don't suspect we will special case daemon set pods yet until we see how that other proposal finishes up.

derekwaynecarr commented:
@vishh - PTAL, made a couple updates based on your feedback.

* removed the --eviction-strategy flag exposed to system admin since we will only support one for now, the moment we find a need to differentiate, we can add a flag.  I agree now that we may never need one depending on how the priority work works out.
* removed references to other resources, i.e. cpu to avoid confusion.
* added some text on how long we will block to observe a pod termination before selecting an additional pod to evict in case of stuck processes.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,219 @@
+`<eviction-signal><operator><quantity>`
+
+* valid `eviction-signal` tokens are `memory.available`
+* valid `operator` tokens are `<`
```
Are we going to support `>` as well? IIRC, your recent PR did have support for `>`.

derekwaynecarr replied:
The PR had it because I was experimenting with cpu usage awhile back and forgot to take it out.  I can remove support for that operator in that PR.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,219 @@
+The `kubelet` will respect graceful termination periods for pods when
+doing eviction.  If a pod is not terminated within its graceful termination
+period because a container does not happen to die (i.e. processes stuck in
+disk IO for example), the `kubelet` may select an additional pod to fail.
```
Considering that the pod would be terminated abruptly by the kernel in the event of a system OOM, should we support graceful terminations here?

The amount of memory that can be sacrificed for the sake of stability will be pretty low in reality. If we were to wait for >30s to evict a pod, system OOM will most likely kick in, making this whole eviction feature mute.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,219 @@
+best to not evict a DB if it requires a local SSD
+* target pods with lower warm-up times before they are ready to accept traffic
+since they would be ready quicker on other nodes
+* etc.
```
I wonder if these policies will give unfair advantages to certain applications... What if a best effort pod chooses to use local storage and has a large warm-up time?

derekwaynecarr replied:
I am happy to remove everything under "Future iterations"

vishh commented on docs/proposals/kubelet-eviction.md:
````diff
@@ -0,0 +1,219 @@
+--eviction-soft-grace-period="memory.available=30s"
+```
+
+If the `kubelet` sees that it has less than `256Mi` of memory available
````
The scheduler will soon support usage based scheduling. Whatever logic that kubelet follows should be shared with the scheduler as well ideally. 
If the node is partitioned statically, this section effectively reduces the amount of `Available` resources right?

vishh replied:
Answering myself: Node Conditions will be used to make the nodes under memory pressure least preferable.

derekwaynecarr replied:
I can add a blurb in this section that the scheduler should use `MemoryPressure` to dissuade placing additional best effort pods on the node.

vishh commented:
@derekwaynecarr just did another review pass.

derekwaynecarr commented:
Per discussion in sig-node slack:

1. Add a `MemoryPressure` node condition when eviction thresholds are met
2. Clarify that hard eviction thresholds always require grace period = 0
3. Add ability to define a max soft eviction pod termination grace period, soft eviction thresholds when met will use the min(max soft eviction pod termination grace period, pod grace period)
4. Runtime interface needs to allow for kill pod to take options to override grace period

derekwaynecarr commented:
@vishh - updates made as requested, PTAL

vishh commented on docs/proposals/kubelet-eviction.md:
````diff
@@ -0,0 +1,259 @@
+```
+--eviction-soft="": A set of eviction thresholds (e.g. memory.available<1.5Gi) that if met over a corresponding grace period would trigger a pod eviction.
+--eviction-soft-grace-period="": A set of eviction grace periods (e.g. memory.available=1m30s) that correspond to how long a soft eviction threshold must hold before triggering a pod eviction.
+--eviction-soft-max-pod-termination-grace-period-seconds="0": Maximum allowed pod termination grace period to use when evicting pods from the node in response to a soft eviction threshold being met.
````
This is a longgggg flag name. Can we make it shorter and instead rely on the description to provide more meaning?

derekwaynecarr replied:
I struggled with naming here.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,259 @@
+A hard eviction threshold has no grace period, and if observed, the `kubelet`
+will take immediate action to reclaim the associated starved resource.  If a
+hard eviction threshold is met, the `kubelet` will kill the pod immediately
+and not respect its termination grace period.
```
Alternatively, `kubelet will kill the pod immediately with no graceful termination`

vishh commented on docs/proposals/kubelet-eviction.md:
````diff
@@ -0,0 +1,259 @@
+To configure hard eviction thresholds, the following flag will be supported:
+
+```
+--eviction-hard="": A set of eviction thresholds (e.g. memory.available<1Gi) that if met would trigger a pod eviction.
````
Would it make sense to support %age here? How do you envision admins setting this flag? Is a homogeneous cluster assumed here?

derekwaynecarr replied:
I think %age is confusing to some degree to express in a flag as it makes the syntax more complex, and I am not sure how it helps in a heterogenous cluster versus homogenous cluster.  We don't specify system-reserved or kube-reserved in relative percentages, so I would prefer to be consistent here.

derekwaynecarr replied:
>How do you envision admins setting this flag?

Based on workload and experience.  Average number of pods per node, etc.  Many clusters will have utilization metrics that they have seen and will have experienced spikes in utilization that induced OOM.  If utilization is a top priority, then I suspect the flag would be set uniformly across the cluster with a value that satisfies their target balanced with their willingness to accept OOM, speed to respond, and ability to procure new nodes.

Many operators are not pushing utilization to huge numbers and are interested in node stability and just the consistent control plane afforded by Kubernetes/OpenShift.  In those cases, I suspect they would set eviction thresholds to happen earlier since they probably have slack on other nodes in the cluster and are comfortable or desire lower utilization for their workload.

For the vast majority though, I would imagine folks will follow this pattern:

* operators are not seeing OOMs, so they set nothing
* cluster gets popular, users do more, start to experience OOM on nodes regularly
* operator files a bug with us and buys more nodes temporarily
* operators notice users are not doing the right thing with their requests, or using best effort pods to run dbs, etc.
* operators decide to experiment on a month-by-month basis to set eviction flags until their incidence of oom observed reaches satisfactory outcomes.
* users continue to do unexpected things like run dbs in besteffort pods

>Is a homogeneous cluster assumed here?

No.  Tailoring values based on node profiles is probably something I would expect Ansible to do today probably similar to how we can specify node labels.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,259 @@
+ * block until pod is terminated on node
+
+If a pod is not terminated because a container does not happen to die
+(i.e. processes stuck in disk IO for example), the `kubelet` may select
```
At what point will kubelet give up on a pod termination? Is it a timeout or will it look at the reason for termination failure?

derekwaynecarr replied:
Ideally, we will call `KillPod` directly using the grace period override specified here in #24843 which is a blocking call.

If it returns an error, I think its in our interest to pick another rather than get fancier.

vishh commented on docs/proposals/kubelet-eviction.md:
```diff
@@ -0,0 +1,259 @@
+the strategy targets the largest consumer of the starved resource.
+
+In practice, `Guaranteed` pods should rarely be targeted for eviction unless
+a system daemon (`kubelet`, `docker`, etc.) is using greater than the specified
```
Shouldn't these daemons be restricted to not exceed `system-reserved`? I'd prefer not evicting pods that use less than their request since it translates to the system screwing up.

derekwaynecarr replied:
They should be restricted to not exceed `system-reserved` but they are not, and we ourselves are not limiting or providing a recommendation for `kube-reserved` yet, so I think we need to keep this escape hatch for now.  If I update the text to state that in the future, if a node enforces `system-reserved` and `kube-reserved` limits via a process not yet defined, then we will never evict `Guaranteed` pods?  For now, my bias to evict one `Guaranteed` pod to protect another if I must.

/cc @dchen1107 - please weigh in here, I see no other option if we want to protect node stability given our current enforcement.

derekwaynecarr commented:
@vishh - I updated the flag name, and added some clarifications to the text around scheduler behavior, and kill pod error checking.  I disagree with the expectation that `Guaranteed` pods should never be evicted since we do not yet have a foundation in place to support that claim, but I am open to being convinced because `Guaranteed` pods are not my top concern when thinking about users that will get value out of this feature.

derekwaynecarr commented:
@k8s-bot test this issue https://github.com/kubernetes/kubernetes/issues/24538
